lib/features/pokemon_overview/pokemon_overview_page.dart

```diff
@@ -1,45 +1,106 @@
+
+  @override
+  State<PokemonOverviewPage> createState() => _PokemonOverviewPageState();
+}
+
+class _PokemonOverviewPageState extends State<PokemonOverviewPage> {
+  late TextEditingController _searchController;
+  late bool isSearching;
+  Timer? _debounceTimer;
+
+  @override
+  void initState() {
+    super.initState();
+    isSearching = false;
+    _searchController = TextEditingController();
+  }
+
+  @override
+  void dispose() {
+    _clearSearchedPokemonsOnDispose();
+    _searchController.dispose();
+    _debounceTimer?.cancel();
+    super.dispose();
+  }
 
   @override
   Widget build(BuildContext context) {
     return Scaffold(
       appBar: AppBar(title: const Text(appName)),
-      body: Center(
-        child: pokemons.when(
-          (data) => GridView.builder(
-            itemCount: data.length,
-            gridDelegate: const SliverGridDelegateWithFixedCrossAxisCount(
-              crossAxisCount: gridRowCount,
-              childAspectRatio: cardRatio,
+      body: Container(
+        padding: const EdgeInsets.all(10.0),
+        child: Column(
+          children: [
+            SearchBar(
+              searchController: _searchController,
+              onClear: _clearSearchedPokemons,
+              onChanged: (val) => _onSearchTextChanged(val),
+            ),
+            const VerticalSpace(height: 20.0),
+            Expanded(
+              child: isSearching
+                  ? PokemonGridView(pokemons: widget.searchedPokemons)
+                  : widget.pokemons.when(
+                      (data) => PokemonGridView(pokemons: data),
+                      loading: () => const Center(child: CircularProgressIndicator()),
+                      error: (errorMessage) {
+                        WidgetsBinding.instance.addPostFrameCallback(
+                            (_) => _showErrorMessageSnackbar(context, errorMessage ?? emptyString));
+                        return const BodySmallText(text: errorMessageDefault);
+                      },
+                    ),
             ),
-            itemBuilder: (_, int index) {
-              final pokemon = data[index];
-              return PokemonCard(pokemon: pokemon);
-            },
-          ),
-          loading: () => const CircularProgressIndicator(),
-          error: (errorMessage) {
-            WidgetsBinding.instance
-                .addPostFrameCallback((_) => _showErrorMessageSnackbar(context, errorMessage ?? emptyString));
-            return const Text(errorMessageDefault);
-          },
+          ],
         ),
       ),
     );
   }
+
+  void _clearSearchedPokemons() {
+    widget.clearSearchedPokemons;
+    setState(() {
+      _searchController.text = emptyString;
+      isSearching = false;
+    });
+  }
+
+  void _clearSearchedPokemonsOnDispose() {
+    if (widget.searchedPokemons.isNotEmpty) widget.clearSearchedPokemons;
+  }
+
+  void _onSearchTextChanged(String queryText) {
+    _debounceTimer = Timer(
```

Comment:
seems you need to have this validation above this line to make sure there's only one active debouncer running 🙂

if (_debounceTimer?.isActive ?? false) _debounce?.cancel();